### PR TITLE
Drops PHP 7.3 and Checks for the Intl Extension

### PR DESCRIPTION
**Pull request by gsteel**
|    Q          |   A
|-------------- | ------
| Bugfix        | kindof
| BC Break      | no
| QA            | yes

### Description

Sorry - this is quite large - Initially, I intended to take a look and see what can be done to decouple this from view, but there's quite a lot of tidying up to do first.

Initially, I dropped 7.3 and updated the lock on 7.4 - I also noticed that `intl` is declared in composer so I went through the component and removed all the conditionals that check for the extension.

I discovered that tests were failing for me in unexpected ways that was due to the default locale getting changed and not correctly restored/preserved between tests, so ended up making an abstract parent test case for i18n that preserves and restores the default locale on setUp() and tearDown()

I've also set phpunit to convert deprecations to exceptions and added the prophecy trait to dev deps to squash the deprecations from PHPUnit 9

<del>Also, CS passes locally, but code sniffer itself is chucking out deprecations on 7.4, so we need to upgrade LCS to whatever is current. I can do that here if it suits?</del>

Updated LCS to 2.3 and fixed all the violations 😬

… Also … a bit naughty, but I updated the workflows for renovate bot and set platform to 7.4

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -16,7 +16,13 @@
         "forum": "https://discourse.laminas.dev"
     },
     "config": {
-        "sort-packages": true
+        "sort-packages": true,
+        "platform": {
+            "php": "7.4"
```
Comment (Ocramius) [inline composer.json]: This means `7.4.0`, and `7.4.0` had a bunch of bugs in it :D 

IMO `7.4.99` or such

File: src/Filter/Alpha.php [hunk 2]
```diff
@@ -1,10 +1,17 @@
 <?php
 
+declare(strict_types=1);
```
Comment (Ocramius) [inline src/Filter/Alpha.php]: Won't this crash on scalar integers, for example?

File: src/Translator/Loader/Gettext.php [hunk 3]
```diff
@@ -136,7 +150,7 @@ public function load($locale, $filename)
             $rawHeaders = explode("\n", trim($textDomain['']));
 
             foreach ($rawHeaders as $rawHeader) {
-                list($header, $content) = explode(':', $rawHeader, 2);
+                [$header, $content] = explode(':', $rawHeader, 2);
```
Comment (Ocramius) [inline src/Translator/Loader/Gettext.php:151]: Is `$content` guaranteed to exist?

File: src/Translator/Loader/PhpMemoryArray.php [hunk 4]
```diff
@@ -1,21 +1,26 @@
...
  */
 class PhpMemoryArray implements RemoteLoaderInterface
 {
-    /**
-     * @var array
-     */
+    /** @var array */
     protected $messages;
 
+    /** @param array $messages */
```
Comment (Ocramius) [inline src/Translator/Loader/PhpMemoryArray.php:21]: Could we be more precise about the type of `$messages`, perhaps?

File: src/Translator/Plural/Parser.php [hunk 5]
```diff
@@ -219,7 +224,7 @@ protected function registerSymbol($id, $leftBindingPower = 0)
      */
     protected function getSymbol($id)
     {
-        if (! isset($this->symbolTable[$id])) {
+        if (! isset($this->symbolTable[$id])) { // phpcs:ignore
```
Comment (Ocramius) [inline src/Translator/Plural/Parser.php:225]: Wait what?

File: src/Translator/Plural/Parser.php [hunk 6]
```diff
@@ -320,7 +325,7 @@ protected function getNextToken()
                 if ($this->string[$this->currentPos] === $result) {
                     $this->currentPos++;
                     $id = $result . $result;
-                } else {
+                } else { // phpcs:ignore
```
Comment (Ocramius) [inline src/Translator/Plural/Parser.php:326]: What's going on here? :|

File: src/Validator/Alnum.php [hunk 7]
```diff
@@ -100,7 +108,7 @@ public function isValid($value)
 
         static::$filter->setAllowWhiteSpace($this->options['allowWhiteSpace']);
 
-        if ($value != static::$filter->filter($value)) {
+        if ($value != static::$filter->filter($value)) { // phpcs:ignore
```
Comment (Ocramius) [inline src/Validator/Alnum.php:109]: Not sure why this needs to be ignored

File: src/View/Helper/DateFormat.php [hunk 8]
```diff
@@ -123,7 +116,7 @@ public function setTimezone($timezone)
         $this->timezone = (string) $timezone;
 
         // The method setTimeZoneId is deprecated as of PHP 5.5.0
-        $setTimeZoneMethodName = (PHP_VERSION_ID < 50500) ? 'setTimeZoneId' : 'setTimeZone';
+        $setTimeZoneMethodName = PHP_VERSION_ID < 50500 ? 'setTimeZoneId' : 'setTimeZone';
```
Comment (Ocramius) [inline src/View/Helper/DateFormat.php]: This `PHP_VERSION_ID` check can be dropped

**Review by Ocramius — CHANGES_REQUESTED** [review]
I went half-way with the review: the diff is too risky, in my opinion.

We would first need #37 to be part of this repo, and then we would need to hunt for clear type violations, before merging this.

As it currently is, the addition of `declare(strict_types=1)` on this codebase is too much.

It is also possible to just strip out that one CS rule for now: that would make the change much safer to push through.

File: src/Validator/Alnum.php (hunk 7 above)
Comment (gsteel) [reply src/Validator/Alnum.php:109]: Non-strict comparison

File: src/Translator/Plural/Parser.php (hunk 6 above)
Comment (gsteel) [reply src/Translator/Plural/Parser.php:326]: I didn't want to touch existing source in this pull if I could avoid it: This one is an empty else

File: src/Translator/Plural/Parser.php (hunk 5 above)
Comment (gsteel) [reply src/Translator/Plural/Parser.php:225]: I didn't want to touch existing source in this pull if I could avoid it: This one is an `empty if`

**Comment by gsteel** [comment]
> I went half-way with the review: the diff is too risky, in my opinion.
> 
> We would first need #37 to be part of this repo, and then we would need to hunt for clear type violations, before merging this.
> 
> As it currently is, the addition of `declare(strict_types=1)` on this codebase is too much.
> 
> It is also possible to just strip out that one CS rule for now: that would make the change much safer to push through.

Started looking at the review prior to seeing your comment… Understood. It was pretty cavalier to put so many changes into one pull…

I'll send in a pull for psalm to replace #37 as it looks like it might be dead and see what I can resurrect from here later on…

**Comment by Ocramius** [comment]
TBH, if you send in this patch without `declare(strict_types=1)`, that's already quite OK IMO

**Comment by gsteel** [comment]
@Ocramius - #71 just landed - if that's OK to merge, I'll rebase this on that, update the baseline and strip out the strict types then?

File: src/Translator/Loader/PhpMemoryArray.php (hunk 4 above)
Comment (gsteel) [reply src/Translator/Loader/PhpMemoryArray.php:21]: Correctly annotating messages might be a job for another day - I believe it's a shape that is returned by all the "loaders" and has a very specific shape that would benefit from a type on an interface somewhere…

File: src/Translator/Loader/Gettext.php (hunk 3 above)
Comment (gsteel) [reply src/Translator/Loader/Gettext.php:151]: I believe that gettext headers are `key:maybe-value` pairs - shall I revert this line? tests still pass and that line _is_ covered

Comment (Ocramius) [reply src/Translator/Loader/Gettext.php:151]: The reason why I commented on this is that `declare(strict_types=1)` leads to a potential `$content = null` being passed to a function later on in this method.

**Comment by gsteel** [comment]
@Ocramius and @froschdesign - I've reverted strict_types in src/ and rebased on the psalm merge.

Psalm errors are way down but some of the data providers were making my eyes bleed so I've glossed over a bit on the baseline. I've also addressed the previous review comments where possible.

File: src/Translator/Loader/Gettext.php (hunk 3 above)
Comment (Ocramius) [reply src/Translator/Loader/Gettext.php:151]: No longer applicable, since `declare(strict_types=1)` is gone.

File: src/Validator/DateTime.php [hunk 9]
```diff
@@ -312,7 +298,7 @@ protected function getIntlDateFormatter()
                 $this->getTimeType(),
                 $this->timezone,
                 $this->calendar,
-                $this->pattern
+                $this->pattern ?? ''
```
Comment (Ocramius) [inline src/Validator/DateTime.php:301]: Is an empty pattern acceptable?

**Review by Ocramius — APPROVED** [review]

File: src/Validator/DateTime.php (hunk 9 above)
Comment (Ocramius) [reply src/Validator/DateTime.php:301]: @gsteel only this comment left, before merge :)

Comment (gsteel) [reply src/Validator/DateTime.php:301]: I coalesced this to an empty string from null because strict_types caused errors. The existing tests already cover `null` but I've added a couple of extra tests to make sure that an empty string has the same behaviour as a null.

**Review by Ocramius — APPROVED** [review]
Awesome, thanks @gsteel!
